pkg/etcdcli/cached_members.go

```diff
@@ -0,0 +1,129 @@
+package etcdcli
+
+import (
+	"context"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"sync"
+	"time"
+)
+
+const refreshPeriod = time.Minute
+
+type CachedMembers struct {
+	client  EtcdClient
+	lock    sync.Mutex
+	members memberHealth
+
+	lastRefresh time.Time
+}
+
+func NewMemberCache(client EtcdClient) AllMemberLister {
+	return &CachedMembers{
+		client:      client,
+		lock:        sync.Mutex{},
+		lastRefresh: time.UnixMilli(0),
+	}
+}
+
+func (c *CachedMembers) shouldRefresh() bool {
+	return c.lastRefresh.Add(refreshPeriod).Before(time.Now())
+}
+
+func (c *CachedMembers) refresh(ctx context.Context) error {
+	c.lock.Lock()
```

Comment:
@tjungblu i think we should lock and ulock if `shouldRefresh()` is true 

wdyt ?

Reply:
we always need to lock, because reading the lastRefresh needs a memory barrier

Reply:
i cant get this ? .. what should i change ?

i am working on testing now 👍🏽

Reply:
I'll send you a suggestion on this PR